### ISSUE TITLE
masurca: Problems installing MaSuRCA and GLIBC using Linuxbrew

### ISSUE THREAD

**Issue by philippbayer** [Homebrew/homebrew-science#3385]
As discussed on Twitter in this [thread](https://twitter.com/sjackman/status/704025671845040128)

Currently the installation of MaSuRCA fails with ([entire log here](https://gist.github.com/philippbayer/3c83b7131b3433e2c1e2))

```
cd /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/obj; \
          /gpfs1/homes/pbayer/.linuxbrew/bin/g++-5 -o /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/bin/fragmentDepth /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/obj/fragmentDepth.o /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib/libCA.a   -L/tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib  -L/tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib  -L/usr/lib64  -L/usr/X11R6/lib64 -L/gpfs1/homes/pbayer/.linuxbrew/lib -Wl,--dynamic-linker=/gpfs1/homes/pbayer/.linuxbrew/lib/ld.so -Wl,-rpath,/gpfs1/homes/pbayer/.linuxbrew/lib -pthread -lm -Wl,-O1 -pthread -lm -Wl,-O1
++++++++++++++++++++  markUniqueUnique ++++++++++++++++++++++
cd /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/obj; \
          /gpfs1/homes/pbayer/.linuxbrew/bin/g++-5 -o /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/bin/markUniqueUnique /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/obj/markUniqueUnique.o /tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib/libCA.a   -L/tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib  -L/tmp/masurca20160229-8746-1mt8ybf/MaSuRCA-2.3.2b/CA/Linux-amd64/lib  -L/usr/lib64  -L/usr/X11R6/lib64 -L/gpfs1/homes/pbayer/.linuxbrew/lib -Wl,--dynamic-linker=/gpfs1/homes/pbayer/.linuxbrew/lib/ld.so -Wl,-rpath,/gpfs1/homes/pbayer/.linuxbrew/lib -pthread -lm -Wl,-O1 -pthread -lm -Wl,-O1
/gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so: undefined reference to `__cxa_thread_atexit_impl@GLIBC_2.18'
/gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so: undefined reference to `clock_gettime@GLIBC_2.17'
collect2: error: ld returned 1 exit status
make[1]: *** [fragmentDepth] Error 1
make[1]: *** Waiting for unfinished jobs....
/gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so: undefined reference to `__cxa_thread_atexit_impl@GLIBC_2.18'
/gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so: undefined reference to `clock_gettime@GLIBC_2.17'
collect2: error: ld returned 1 exit status
make[1]: *** [markUniqueUnique] Error 1
make: *** [progs] Error 1
```

As @cjfields [noted](https://twitter.com/cjfields/status/704102716314783744), this seems to be a problem with the bundled Celera Assembler (CA/Linux-amd64/lib/libCA.a) assuming an older GLIBC version which Brew doesn't use (Brew's installed GLIBC is v2.19).

It doesn't matter whether I do `brew install masurca` or manually run setup.sh in the downloaded sources, it's the same error with different paths.

Work-around: Use provided binaries (doesn't work on all servers)

Tried:

0. `brew update` & `brew doctor` (no errors except `Warning: "config" scripts exist outside your system or Homebrew directories` from cluster folders)
1. `brew patchelf masurca` - got: `Error: Formula not installed or up-to-date: homebrew/science/masurca`
2. Manually installed patchelf, worked fine

**Comment by sjackman** [comment]
Please report the output of either `brew config` or `brew gist-logs masurca`.
It looks to me as though the error may be with `libstdc++.so` linking against the wrong version of `glibc`. It looks as though it's picking up your older system `glibc`. Please report
```sh
ldd /gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so
readelf -d /gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so
```

**Comment by philippbayer** [comment]
Sure!

[brew config output](https://gist.github.com/philippbayer/bc4c10ffe807c13e4175)

```
ldd /gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so
	linux-vdso.so.1 (0x00007fffe19ff000)
	libm.so.6 => /gpfs1/homes/pbayer/.linuxbrew/lib/libm.so.6 (0x00002b2820ba4000)
	libc.so.6 => /gpfs1/homes/pbayer/.linuxbrew/lib/libc.so.6 (0x00002b2820e96000)
	/gpfs1/homes/pbayer/.linuxbrew/Cellar/glibc/2.19/lib64/ld-linux-x86-64.so.2 (0x00002b282060e000)
	libgcc_s.so.1 => /gpfs1/homes/pbayer/.linuxbrew/lib/libgcc_s.so.1 (0x00002b2821213000)
```

```
readelf -d /gpfs1/homes/pbayer/.linuxbrew/lib/libstdc++.so

Dynamic section at offset 0x12c130 contains 31 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x000000000000000e (SONAME)             Library soname: [libstdc++.so.6]
 0x000000000000001d (RUNPATH)            Library runpath: [/gpfs1/homes/pbayer/.linuxbrew/lib]
 0x000000000000000c (INIT)               0x893d8
 0x000000000000000d (FINI)               0xeb7a0
 0x0000000000000019 (INIT_ARRAY)         0x323988
 0x000000000000001b (INIT_ARRAYSZ)       88 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x3239e0
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x0000000000000004 (HASH)               0x1c8
 0x0000000000000005 (STRTAB)             0x3341c0
 0x0000000000000006 (SYMTAB)             0x76b0
 0x000000000000000a (STRSZ)              263044 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x32d000
 0x0000000000000002 (PLTRELSZ)           36624 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x804c8
 0x0000000000000007 (RELA)               0x6a7f0
 0x0000000000000008 (RELASZ)             89304 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffc (VERDEF)             0x6a248
 0x000000006ffffffd (VERDEFNUM)          35
 0x000000006ffffffe (VERNEED)            0x6a710
 0x000000006fffffff (VERNEEDNUM)         4
 0x000000006ffffff0 (VERSYM)             0x677ca
 0x000000006ffffff9 (RELACOUNT)          739
 0x0000000000000000 (NULL)               0x0
```

**Comment by sjackman** [comment]
I should have assumed the simpler problem first. The issue is that the `-L/usr/lib64` option coming before the `-L/gpfs1/homes/pbayer/.linuxbrew/lib` option in the link line. The `Makefile` needs to be hacked up to remove `-L/usr/lib64`, which should not be mentioned explicitly, because it's an implicit search directory. And if it's disabled, as in standalone Linuxbrew, it's probably for a good reason.

**Comment by philippbayer** [comment]
Interesting!

I can confirm that if I change
```  
  ifeq ($(MACHINETYPE), amd64)
    ARCH_CFLAGS   += -m64 -DX86_GCC_LINUX
    ARCH_LIB       = /usr/lib64 /usr/X11R6/lib64
  endif
```
to

```  
  ifeq ($(MACHINETYPE), amd64)
    ARCH_CFLAGS   += -m64 -DX86_GCC_LINUX
    ARCH_LIB       = # /usr/lib64 /usr/X11R6/lib64
  endif
```

in the file `CA/src/c_make.as` it compiles fine (this system doesn't have /usr/X11R6 anyway).

**Comment by sjackman** [comment]
Great. Would you like to send a pull request? Something like
```ruby
# Fix libstdc++.so: undefined reference to `clock_gettime@GLIBC_2.17'
inreplace "CA/src/c_make.as", "ARCH_LIB *= /usr/lib64 /usr/X11R6/lib64", ""
```

**Comment by philippbayer** [comment]
Sure! I'll have a look how it works "around here" and send you one in the next few days

**Comment by sjackman** [comment]
https://travis-ci.org/Homebrew/homebrew-science/jobs/113302964#L876
```
############################## AS_OVL ##############################
make[1]: *** No rule to make target `jellyfish/jellyfish.hpp', needed by `AS_OVL_overlap.o'.  Stop.
make: *** [objs] Error 1
```
I'm not sure why this error should come up now and not before. It doesn't seem related to this patch. It might require `ENV.deparallelize` to fix. In any case, I'm going to merge this patch without updating bottles, as it doesn't affect Mac.

**Comment by sjackman** [comment]
Fixed by #3397.
